### PR TITLE
Part 3 - Async Compaction - Track fileIds with pending compaction in FileSystemView to provide correct API semantics

### PR DESCRIPTION
**Pull request by bvaradar**
Track fileIds with pending compaction in FileSystemView to provide correct API semantics

### CODE REVIEW

**Comment by bvaradar** [comment]
@n3nash : This PR takes care of correctly tracking fileIds which have pending (inflight + requested) compactions in file-system view to provide correct functionality in FileSystemView APIs.

**Comment by bvaradar** [comment]
@n3nash : ready for review

cc @vinothchandar

File: hoodie-common/src/main/java/com/uber/hoodie/common/table/view/HoodieTableFileSystemView.java [hunk 1]
```diff
@@ -79,10 +88,33 @@ public HoodieTableFileSystemView(HoodieTableMetaClient metaClient,
     this.visibleActiveTimeline = visibleActiveTimeline;
     this.fileGroupMap = new HashMap<>();
     this.partitionToFileGroupsMap = new HashMap<>();
-    //TODO: vb Will be implemented in next PR
-    this.fileIdToPendingCompactionInstantTime = new HashMap<>();
-  }
 
+    // Build fileId to Pending Compaction Instants
+    List<HoodieInstant> pendingCompactionInstants =
```
Comment (n3nash) [inline hoodie-common/src/main/java/com/uber/hoodie/common/table/view/HoodieTableFileSystemView.java:90]: @bvaradar I saw this in the other diff too, any overlap here ?

File: hoodie-common/src/test/java/com/uber/hoodie/common/table/view/HoodieTableFileSystemViewTest.java [hunk 2]
```diff
@@ -33,16 +34,20 @@
...
 import java.io.IOException;
+import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
 import java.util.Set;
 import java.util.UUID;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
+import javafx.util.Pair;
```
Comment (n3nash) [inline hoodie-common/src/test/java/com/uber/hoodie/common/table/view/HoodieTableFileSystemViewTest.java]: use apache commons Pair ?

File: hoodie-common/src/main/java/com/uber/hoodie/common/table/view/HoodieTableFileSystemView.java [hunk 3]
```diff
@@ -79,10 +88,33 @@ public HoodieTableFileSystemView(HoodieTableMetaClient metaClient,
...
+        metaClient.getActiveTimeline().filterPendingCompactionTimeline().getInstants().collect(Collectors.toList());
+    Map<String, String> fileIdToPendingCompactionMap = new HashMap<>();
+    pendingCompactionInstants.stream().flatMap(instant -> {
+      try {
+        HoodieCompactionWorkload workload = AvroUtils.deserializeHoodieCompactionWorkload(
```
Comment (n3nash) [inline hoodie-common/src/main/java/com/uber/hoodie/common/table/view/HoodieTableFileSystemView.java]: Rename to AvroUtils.convertCompactionWorkload() or AvroUtils.deserializeCompactionWorkload() ? Naming seems too long.

File: hoodie-common/src/main/java/com/uber/hoodie/common/table/view/HoodieTableFileSystemView.java (hunk 1 above)
Comment (bvaradar) [reply hoodie-common/src/main/java/com/uber/hoodie/common/table/view/HoodieTableFileSystemView.java:90]: This PR will be applied on top of the other PR

File: hoodie-common/src/test/java/com/uber/hoodie/common/table/view/HoodieTableFileSystemViewTest.java (hunk 2 above)
Comment (bvaradar) [reply hoodie-common/src/test/java/com/uber/hoodie/common/table/view/HoodieTableFileSystemViewTest.java]: Done

File: hoodie-common/src/main/java/com/uber/hoodie/common/table/view/HoodieTableFileSystemView.java (hunk 3 above)
Comment (bvaradar) [reply hoodie-common/src/main/java/com/uber/hoodie/common/table/view/HoodieTableFileSystemView.java]: Done

**Comment by bvaradar** [comment]
@n3nash: Addressed review comments. Please take one more pass when you get a chance.

File: hoodie-common/src/main/java/com/uber/hoodie/common/util/CompactionUtils.java [hunk 4]
```diff
@@ -97,4 +102,53 @@ public static HoodieCompactionOperation buildHoodieCompactionOperation(Compactio
   public static CompactionOperation buildCompactionOperation(HoodieCompactionOperation hc) {
     return CompactionOperation.convertFromAvroRecordInstance(hc);
   }
+
+  /**
+   * Get all pending compaction workloads along with their instants
+   *
+   * @param metaClient Hoodie Meta Client
+   */
+  public static List<Pair<HoodieInstant, HoodieCompactionWorkload>> getAllPendingCompactionWorkloads(
```
Comment (vinothchandar) [inline hoodie-common/src/main/java/com/uber/hoodie/common/util/CompactionUtils.java]: is there a test for this file? I think we should add tons of tests around these methods in this file

File: hoodie-common/src/main/java/com/uber/hoodie/common/util/CompactionUtils.java [hunk 5]
```diff
@@ -97,4 +102,53 @@ public static HoodieCompactionOperation buildHoodieCompactionOperation(Compactio
...
+      HoodieTableMetaClient metaClient) {
+    List<HoodieInstant> pendingCompactionInstants =
+        metaClient.getActiveTimeline().filterPendingCompactionTimeline().getInstants().collect(Collectors.toList());
+    return pendingCompactionInstants.stream().map(instant -> {
+      try {
+        HoodieCompactionWorkload workload = AvroUtils.deserializeCompactionWorkload(
+            metaClient.getActiveTimeline().getInstantDetails(instant).get());
```
Comment (vinothchandar) [inline hoodie-common/src/main/java/com/uber/hoodie/common/util/CompactionUtils.java]: this is an additional FS open + read.. We should seriously think about immediately priotizitng the filesytemview caching after this, to limit the namenode calls

**Review by vinothchandar — COMMENTED** [review]
LGTM overall

File: hoodie-common/src/main/java/com/uber/hoodie/common/util/CompactionUtils.java (hunk 4 above)
Comment (bvaradar) [reply hoodie-common/src/main/java/com/uber/hoodie/common/util/CompactionUtils.java]: Unit tests added in TestCompactionUtils

File: hoodie-common/src/main/java/com/uber/hoodie/common/util/CompactionUtils.java (hunk 5 above)
Comment (n3nash) [reply hoodie-common/src/main/java/com/uber/hoodie/common/util/CompactionUtils.java]: +1 @bvaradar could you please open a ticket for this ?

Comment (bvaradar) [reply hoodie-common/src/main/java/com/uber/hoodie/common/util/CompactionUtils.java]: https://github.com/uber/hudi/issues/433

**Comment by bvaradar** [comment]
Committed as part of Consolidated diff

**Pull request closed without merging**